src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -74,34 +172,57 @@ static void WriteAsset(ObjectWriter writer, ISerializerService serializer, Solut
+
+                    // Now buffer in the rest of the data we need to read.  Because we're reading as much data in as
+                    // we'll need to consume, all further reading (for this single item) can handle synchronously
+                    // without worrying about this blocking the reading thread on cross-process pipe io.
+                    var fillReadResult = await pipeReader.ReadAtLeastAsync(length, cancellationToken).ConfigureAwait(false);
+
+                    // Note: we have let the pipe reader know that we're done with 'read at least' call, but that we
+                    // haven't consumed anything from it yet.  Otherwise it will throw that another read can't start
+                    // from within ObjectReader.GetReader below.
+                    pipeReader.AdvanceTo(fillReadResult.Buffer.Start);
+
+                    // Now do the actual read of the data, synchronously, from the buffers that are now in memory within
+                    // our process.  These reads will move the pipe-reader forward, without causing any blocking on
+                    // async-io.
+                    using var reader = ObjectReader.GetReader(pipeReaderStream, leaveOpen: true, cancellationToken);
+
+                    var checksum = Checksum.ReadFrom(reader);
+                    var kind = (WellKnownSynchronizationKind)reader.ReadInt32();
+
+                    // in service hub, cancellation means simply closed stream
+                    var result = serializerService.Deserialize(kind, reader, cancellationToken);
+                    Contract.ThrowIfNull(result);
+                    callback(checksum, (T)result, arg);
+                }
             }
-        }
 
-        public static void ReadData<T, TArg>(
-            Stream stream, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, Action<int, T, TArg> callback, TArg arg, CancellationToken cancellationToken)
-        {
-            using var reader = ObjectReader.GetReader(stream, leaveOpen: true, cancellationToken);
-
-            // Ensure that no invariants were broken and that both sides of the communication channel are talking about
-            // the same pinned solution.
-            var responseSolutionChecksum = Checksum.ReadFrom(reader);
-            Contract.ThrowIfFalse(solutionChecksum == responseSolutionChecksum);
-
-            for (int i = 0, n = objectCount; i < n; i++)
+            static (byte, int) ReadSentinelAndLength(ReadResult readResult)
             {
-                var kind = (WellKnownSynchronizationKind)reader.ReadInt32();
-
-                // in service hub, cancellation means simply closed stream
-                var result = serializerService.Deserialize(kind, reader, cancellationToken);
-                Contract.ThrowIfNull(result);
-                callback(i, (T)result, arg);
+                var sequenceReader = new SequenceReader<byte>(readResult.Buffer);
```

Comment:
a SequenceReader ensures that if the data overlaps internal buffers, it is represented to us as a contiguous sequence.